Colab-TextGen-GPU.ipynb

```diff
@@ -0,0 +1,144 @@
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# oobabooga/text-generation-webui\n",
+        "\n",
+        "After running both cells, a public gradio URL will appear at the bottom in a few minutes. You can optionally generate an API link.\n",
+        "\n",
+        "* Project page: https://github.com/oobabooga/text-generation-webui\n",
+        "* Gradio server status: https://status.gradio.app/"
+      ],
+      "metadata": {
+        "id": "MFQl6-FjSYtY"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#@title 1. Keep this tab alive to prevent Colab from disconnecting you { display-mode: \"form\" }\n",
+        "\n",
+        "#@markdown Press play on the music player that will appear below:\n",
+        "%%html\n",
+        "<audio src=\"https://oobabooga.github.io/silence.m4a\" controls>"
+      ],
+      "metadata": {
+        "id": "f7TVVj_z4flw"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#@title 2. Launch the web UI\n",
+        "\n",
+        "#@markdown If unsure about the branch, write \"main\" or leave it blank.\n",
+        "\n",
+        "import torch\n",
+        "from pathlib import Path\n",
+        "\n",
+        "if Path.cwd().name != 'text-generation-webui':\n",
+        "  print(\"Installing the webui...\")\n",
+        "\n",
+        "  !git clone https://github.com/oobabooga/text-generation-webui\n",
+        "  %cd text-generation-webui\n",
+        "\n",
+        "  torver = torch.__version__\n",
+        "  print(f\"TORCH: {torver}\")\n",
+        "  is_cuda118 = '+cu118' in torver  # 2.1.0+cu118\n",
+        "  is_cuda117 = '+cu117' in torver  # 2.0.1+cu117\n",
+        "\n",
+        "  textgen_requirements = open('requirements.txt').read().splitlines()\n",
+        "  if is_cuda117:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu117').replace('+cu122', '+cu117').replace('torch2.1', 'torch2.0') for req in textgen_requirements]\n",
+        "  elif is_cuda118:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu118').replace('+cu122', '+cu118') for req in textgen_requirements]\n",
+        "  with open('temp_requirements.txt', 'w') as file:\n",
+        "      file.write('\\n'.join(textgen_requirements))\n",
+        "\n",
+        "  !pip install -r extensions/openai/requirements.txt --upgrade\n",
+        "  !pip install -r temp_requirements.txt --upgrade\n",
+        "\n",
+        "  print(\"\\033[1;32;1m\\n --> If you see a warning about \\\"previously imported packages\\\", just ignore it.\\033[0;37;0m\")\n",
+        "  print(\"\\033[1;32;1m\\n --> There is no need to restart the runtime.\\n\\033[0;37;0m\")\n",
```

Comment:
**Issue**: The approach to modifying 'requirements.txt' based on CUDA version detections is hard-coded and might not be future-proof or scalable with new CUDA versions or requirements changes. <br> **Fix**: Abstract the CUDA version checks and requirements adjustments into a separate, more dynamic method. This method should parse the current environment and adjust requirements in a more flexible manner. <br> **Code Suggestion**: 
 ```
 - if is_cuda117:
 -     textgen_requirements = [req.replace('+cu121', '+cu117').replace('+cu122', '+cu117').replace('torch2.1', 'torch2.0') for req in textgen_requirements]
 - elif is_cuda118:
 -     textgen_requirements = [req.replace('+cu121', '+cu118').replace('+cu122', '+cu118') for req in textgen_requirements]
 + def adjust_requirements_for_cuda(version, requirements):
 +     new_requirements = []
 +     for req in requirements:
 +         if 'torch' in req:
 +             req = req.replace('+cu121', f'+cu{version}').replace('+cu122', f'+cu{version}')
 +         new_requirements.append(req)
 +     return new_requirements
 + textgen_requirements = adjust_requirements_for_cuda(cuda_version, textgen_requirements)
 ```